### PR TITLE
Add `ReduceLROnPlateau` learning rate scheduler

### PR DESCRIPTION
**Pull request by brylee10**
I encountered a need for a learning rate scheduler analogous to the PyTorch `ReduceLROnPlateau` so I rolled my own. I figured others may find it useful as well so made a PR with it here. 

There is an open thread about adding this feature from last year (https://github.com/huggingface/candle/discussions/2224). It looks like there was code written to add a core `Scheduler` trait to set up for learning rate schedulers which was written but never merged (https://github.com/huggingface/candle/pull/2225).  

I think a core trait like that makes sense and I think adding this feature first is still compatible with that design, if it is eventually merged. One note is that `Scheduler` trait from the PR does not take additional arguments in the `step()` function, but `ReduceLROnPlateau` is an exception to the [other PyTorch LR schedulers ](https://github.com/pytorch/pytorch/blob/v2.6.0/torch/optim/lr_scheduler.py#L1363) since it takes a metric as argument. 

For now, `ReduceLROnPlateau` is standalone, but I think the `Scheduler` trait proposed can be generalized to support this, or the API of `ReduceLROnPlateau` could be modified to be consistent: 
```rust
scheduler.set_metric(loss);
scheduler.step();
```

### CODE REVIEW

**Comment by brylee10** [comment]
FYI @EricLBuehler

**Comment by brylee10** [comment]
Got it. PyTorch has a plethora which I understand would be onerous to all support. The motivation was I noticed no other schedulers existed and it seems others had opened discussions about a similar need for at least one of the common schedulers like this. I thought giving one common scheduler would be a useful baseline, similar to how Candle supports SGD and AdamW as two examples of optimizers but stops there leaving the rest to [candle-optimizers](https://docs.rs/candle-optimisers/latest/candle_optimisers/). 

However, I see how it could be out of scope for the main repo given it is a "core" framework, so maybe schedulers can instead be brought into a separate crate.
